### PR TITLE
Improvements to Opta and Wyscout deserializers

### PR DESCRIPTION
**Pull request by DriesDeprest**
Made the following changes / fixes:
Opta:

- Mark events as counter attacks
- Fix pass events without end coordinates
- Introduce shot results

Wyscout
- Handle infractions better
- Start timestamp from 0 again at start of period 2.

### CODE REVIEW

File: kloppy/infra/serializers/event/opta/deserializer.py [hunk 1]
```diff
@@ -226,16 +229,13 @@ def _parse_f24_datetime(dt_str: str) -> float:
...
-            x=float(raw_qualifiers[140]), y=float(raw_qualifiers[141])
-        )
         result = PassResult.COMPLETE
     else:
         result = PassResult.INCOMPLETE
-        # receiver_coordinates = None
-        receiver_coordinates = Point(
-            x=float(raw_qualifiers[140]), y=float(raw_qualifiers[141])
-        )
+    receiver_coordinates = Point(
```
Comment (JanVanHaaren) [inline kloppy/infra/serializers/event/opta/deserializer.py:235]: What is the rationale behind using (0,0) as the default location for the receiver? I'd rather make it explicit to the user that the location of the receiver is unknown. We could perhaps consider implementing the _Null Object_ pattern for the `Point` class to deal with situations where locations are missing.

File: kloppy/infra/serializers/event/wyscout/deserializer_v3.py [hunk 2]
```diff
@@ -184,12 +183,10 @@ def _parse_foul(raw_event: Dict) -> Dict:
 def _parse_card(raw_event: Dict) -> Dict:
     qualifiers = _generic_qualifiers(raw_event)
     card_type = None
-    if _has_tag(raw_event, wyscout_tags.RED_CARD):
-        card_type = CardType.RED
-    elif _has_tag(raw_event, wyscout_tags.YELLOW_CARD):
+    if _check_secondary_event_types(raw_event, ["yellow_card"]):
```
Comment (JanVanHaaren) [inline kloppy/infra/serializers/event/wyscout/deserializer_v3.py:186]: What happens when a player gets a second yellow card?

File: kloppy/infra/serializers/event/wyscout/deserializer_v3.py [hunk 3]
```diff
@@ -401,6 +404,25 @@ def deserialize(self, inputs: WyscoutInputs) -> EventDataset:
...
+                        foul_event_args = _parse_foul(raw_event)
+                        event = self.event_factory.build_foul_committed(
+                            **foul_event_args, **generic_event_args
+                        )
+                        # We already append event to events
+                        # as we potentially have a card and foul event for one raw event
+                        if event and self.should_include_event(event):
+                            events.append(transformer.transform_event(event))
+                        continue
+                    if "yellow_card" in secondary_event_types:
```
Comment (JanVanHaaren) [inline kloppy/infra/serializers/event/wyscout/deserializer_v3.py:418]: How about red cards?

File: kloppy/tests/test_wyscout.py [hunk 4]
```diff
@@ -40,12 +45,3 @@ def test_correct_v2_deserialization(self, event_v2_data: Path):
     def test_correct_auto_recognize_deserialization(self, event_v2_data: Path):
         dataset = wyscout.load(event_data=event_v2_data, coordinates="wyscout")
         assert dataset.records[2].coordinates == Point(29.0, 6.0)
-
-    def test_correct_handling_of_corner_shot(self, event_v3_data: Path):
```
Comment (JanVanHaaren) [inline kloppy/tests/test_wyscout.py:44]: Why has this test been removed?

File: kloppy/infra/serializers/event/wyscout/deserializer_v3.py [hunk 5]
```diff
@@ -347,6 +344,12 @@ def deserialize(self, inputs: WyscoutInputs) -> EventDataset:
                     "period": periods[-1],
                     "timestamp": float(
                         raw_event["second"] + raw_event["minute"] * 60
+                    )
+                    if period_id == 1
+                    else float(
```
Comment (JanVanHaaren) [inline kloppy/infra/serializers/event/wyscout/deserializer_v3.py:349]: How about matches that go into extra time or penalty shoot-outs?

**Review by JanVanHaaren — COMMENTED** [review]
Thanks for the pull request! I've added a few questions to help me better understand the proposed changes.

File: kloppy/infra/serializers/event/opta/deserializer.py (hunk 1 above)
Comment (DriesDeprest) [reply kloppy/infra/serializers/event/opta/deserializer.py:235]: Only one event, in one match in over a thousand analysed Opta F 24 files, had missing receiver coordinates. Therefore, I went for the quick and dirty solution, (0,0) as the default location, that at least allowed loading in the data.

But would of course be fine with the null object pattern for Point class, I'll have a look.

**Comment by DriesDeprest** [comment]
I don't understand how my latest commit could cause the failing checks? 
https://github.com/PySport/kloppy/actions/runs/5321135197/jobs/9635847489?pr=198
@JanVanHaaren do you have an idea what is going on?

**Comment by JanVanHaaren** [comment]
@DriesDeprest I believe that the issue is caused by a regression bug, see https://github.com/actions/setup-python/issues/682. Yesterday's build used Python 3.7.**16**, whereas today's build used Python 3.7.**17**.

**Comment by DriesDeprest** [comment]
> @DriesDeprest I believe that the issue is caused by a regression bug, see [actions/setup-python#682](https://github.com/actions/setup-python/issues/682). Yesterday's build used Python 3.7.**16**, whereas today's build used Python 3.7.**17**.

What do you propose we do with this? Wait until they fix it or pin the Python version 3.7.**16** for now in our build?

**Comment by koenvo** [comment]
Can you merge master in please?

**Comment by DriesDeprest** [comment]
@JanVanHaaren checks are now passing (thx Koen!).
Are your comments resolved with my answers and new commits? :)

File: kloppy/infra/serializers/event/opta/deserializer.py (hunk 1 above)
Comment (JanVanHaaren) [reply kloppy/infra/serializers/event/opta/deserializer.py:235]: I'm fine with sticking to your current approach if missing receiver coordinates are rare.

**Comment by JanVanHaaren** [comment]
@DriesDeprest Thank you for your changes. Two of my questions are still open. It would be good to hear your thoughts.

**Comment by DriesDeprest** [comment]
@JanVanHaaren I've tagged you in the different remaining open questions / conversations with follow-up comments or questions, feel free to mark them as resolved when clear for you or pose additional questions.

**Comment by JanVanHaaren** [comment]
@DriesDeprest Unfortunately, both @koenvo and I can't see your comments and questions.

**Comment by DriesDeprest** [comment]
@JanVanHaaren I made the same mistake as described here: https://github.com/orgs/community/discussions/10369
But should have submitted my review now, and you should now see my comments / questions 😄

File: kloppy/tests/test_wyscout.py (hunk 4 above)
Comment (JanVanHaaren) [reply kloppy/tests/test_wyscout.py:44]: Okay, sounds good!

File: kloppy/infra/serializers/event/wyscout/deserializer_v3.py (hunk 2 above)
Comment (JanVanHaaren) [reply kloppy/infra/serializers/event/wyscout/deserializer_v3.py:186]: Yes, all clear to me now. I had expected that they would differentiate between second yellow cards and straight red cards. Thanks!

File: kloppy/infra/serializers/event/wyscout/deserializer_v3.py (hunk 5 above)
Comment (JanVanHaaren) [reply kloppy/infra/serializers/event/wyscout/deserializer_v3.py:349]: Wyscout has an example in the API documentation. They distinguish between five different periods.

Currently, the deserializer would struggle with matches that go into extra time already due to how the `matchPeriod` field is handled. Hence, I'm fine with keeping your current approach.

```json
{
  "id": 533182244,
  "matchId": 2050345,
  "matchPeriod": "P",
  "minute": 125,
  "second": 45,
  "matchTimestamp": "02:05:45.456",
  "videoTimestamp": "8760.456735",
  "relatedEventId": null,
  "type": {
    "primary": "postmatch_penalty",
    "secondary": []
  },
  "location": {
    "x": 90,
    "y": 50
  },
  "team": {
    "id": 1841,
    "name": "Zenit",
    "formation": "4-1-3-2"
  },
  "opponentTeam": {
    "id": 1843,
    "name": "Lokomotiv Moskva",
    "formation": "4-4-1-1"
  },
  "player": {
    "id": 17686,
    "name": "D. Criscito",
    "position": "LB"
  },
  "pass": null,
  "shot": {
    "bodyPart": "left_foot",
    "isGoal": true,
    "onTarget": true,
    "goalZone": "gbr",
    "xg": null,
    "postShotXg": null,
    "goalkeeperActionId": 533182178,
    "goalkeeper": {
      "id": 8992,
      "name": "Guilherme",
      "reflexSave": false
    }
  },
  "groundDuel": null,
  "aerialDuel": null,
  "infraction": null,
  "carry": null,
  "possession": null
}
```

**Review by JanVanHaaren — APPROVED** [review]
No worries! The pull request looks good to me. Thank you.

**Comment by koenvo** [comment]
It seems there are some conflicts that prevent merging this PR. Can you merge master in?

**Comment by DriesDeprest** [comment]
Done!

**Pull request merged**